### PR TITLE
dbv2 upgrade - alternative approach

### PR DESCRIPTION
**Pull request by neurosis69**
Here is my rewritten dbv2 upgrade script as follow up to the discussion in  #9613.
The main focus was to reduce runtime on low-end hardware (RPi) or having the database on HDDs.

**Alternative upgrade parameters**

```
chia db upgrade_alt -h
Usage: chia db upgrade_alt [OPTIONS]

Options:
  --input PATH            specify input database file
  --output PATH           specify output database file
  --hdd                   select if database is located on HDD device.
                          (DEFAULT=false)

  --offline               select to upgrade database with no active full node.
                          (DEFAULT=false)

  --temp-store-path PATH  specify path used as temporary db destination (Linux
                          only)

  --check-only            specify to check only if enough disk space is
                          available for upgrade. (DEFAULT=false)

  --no-update-config      don't update config file to point to new database.
                          When specifying a custom output file, the config
                          will not be updated regardless

  -h, --help              Show this message and exit.
```

**Runtimes**

The upgrade runtime is reduced up to 50% compared to the original approach.
But this certainly needs some additional tests/comparisons. 

If not other stated, the following upgrade durations are from a database which was manually vacuumed a few days before the upgrade was executed.

System|DB-Device|update time|update_alt time|parameter
--|--|--|--|--
RPi4;8GB|SSD, Samsung 840 evo 1tb|130m30.365s|67m2.213s|offline
RPi4;8GB|SSD, Kingston A400 240GB|234m40.001s|108m57.520s|offline
RPi4;8GB|SSD, Kingston A400 240GB|274m46.893s|125m18.824s|online
i7-11700K;128GB|NVME-SSD, 3 x 1 TB RAID 0|11m40.183s|8m5.468s|offline
i7-11700K;128GB|USB3-HDD, 2 TB|aborted after<br>261m*|15m48.932s|offline, hdd
i7-11700K;128GB|USB3-HDD, 2 TB|n/a|16m9.359s|online, hdd

\* ... at the end only 1866k coins were upgraded with a rate of 89.8 coins/s (avg ~186 coins/s), total of coins is > 43000k. full_blocks migration was finished after ~93m

**The upgrade is split into the following steps**

1. create database_version table and save version
2. create current_peak table and save current peak
3. create empty target v2 tables
  decided to don't source these statements from block_store.py, ... to not load LRU caches or such       
4. convert full_blocks to v2 and load data to target v2 tables
5. create full_blocks indexes on v2 tables
decided to create indexes right after table data migration to get possible use of data in fs cache
6. convert coin_store tables to temp sqlite db file
decided to use this approach and disconnect/recoonect sqlite sessions to get use of mmap
7. convert sub_epoch tables to temp sqlite db file
8. migrate coin_store temp data to target v2 sqlite file
9. create Indexes on coin_store tables
10. migrate sub_epoch tables to target v2 sqlite file
11. convert and migrate hints table to target v2 sqlite file
12. create indexes on hints table
13. finally update the database version to 2

I tried to adress the reasons for the changes in script comments, but here is an overview.

**Change Overview**
- added check_db function to check for sufficient disk space
- added connect_to_db function
- `chia db upgrade` click menu
   - added --check-only menu switch to check without starting upgrade
   - added --hdd menu switch to use slightly different migration approch for probably better performance if db is on HDD dev
   - added --offline switch to start the migration using an offline source database
   - added --temp-store-path to let the user specify a temp directory
- replaced aiosqlite by sqlite3 module
- added psutil module to check for disk usage and virtual memory
- changed migration method to 
   - use `attach database ...` 
   - reconnect after succesfull migration of 1 table to get new mmap
   - use vacuum-like method (`INSERT INTO ... SELECT * FROM ...`) only for certain tables and ignore indexes
   - use cursor arraysize with fetchmany
   - use recursive CTE to migrate main chain path only
   - create indexes directly after table
   - prepared two scenarios, one for migration having less than 3.5 GiB of available memory and one with more
   - for some inserts, added `order by <PRIMARY KEY>` for better insert performance on tables with PK
- prepared page_size change
- prepared mem shrink rate
- moved all table/index definitions out of other files into the main migration script (suboptimal but has less memory footprint)

**Known Issues**
- ETA and rate is not working

**Follow-up Questions**
- What is the table sub_epoch_segments_v3 used for?
   - This table would imo benefit from the largest possible page size of 64k, due to the big size of its rows (on avg ~600k).
   - It would probably be a good idea to move this table into a separate sqlite db file.
   - The data is scattered over the whole sqlite file.
      Statement to check:
```sql
select (pageno - pageno%100000) as modpageno, count(1) as pagecount from dbstat where name = 'sub_epoch_segments_v3' group by modpageno;
```
- Are there plans to change the db page_size for other tables?
   - According to [this information](https://github.com/Chia-Network/chia-blockchain/pull/9613#issuecomment-1013798743) I'd try out a page size of 16k. But this certainly needs to be tested to find the best sizing.

### CODE REVIEW

**Comment by neurosis69** [comment]
@arvidn here is my upgrade approach, please have a look. 
I tried to add it as additional function, so they can be tested side-by-side.

**Comment by mariano54** [comment]
We should only have one upgrade script not two, so if we prefer this, we should remove the other one. This looks like a massive upgrade, thank you @neurosis69.

**Comment by neurosis69** [comment]
Thanks, nice to hear. 
But I can't stress it enough, I have literally no exp with python and even no serious programming exp at all. 
So there may be many bad practice examples and don'ts in this code.

**Review by arvidn — COMMENTED** [review]
I agree that we only should have one conversion function. I think it would make sense to consider some of the things you do here in the main conversion function. It helps review them if you would submit them as separate PRs (or a single PR with somewhat independent commits). Against the existing conversion function.

The `--offline` option doesn't seem to do anything useful. As far as I can tell, it's always safe to turn off journaling mode in this context.

File: chia/cmds/db_upgrade_alt_func.py [hunk 1]
```diff
@@ -0,0 +1,850 @@
...
+    if out_db_path is None:
+        db_path_replaced = db_pattern.replace("CHALLENGE", selected_network).replace("_v1_", "_v2_")
+        out_db_path = path_from_root(root_path, db_path_replaced)
+        mkdir(out_db_path.parent)
+
+    if check_only:
+        check_db(in_db_path, temp_store_path, hdd, check_only)
+    else:
+        # Commented out to save time during tests
+        # check_db(in_db_path, temp_store_path, hdd, check_only)
```
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py:60]: yes, my initial plan was to enforce this check

File: chia/cmds/db_upgrade_alt_func.py [hunk 2]
```diff
@@ -0,0 +1,850 @@
...
+
+    print(f"\n\nLEAVING PREVIOUS DB FILE UNTOUCHED {in_db_path}\n")
+
+
+def check_db(
+    open_in_path: Path,
+    temp_store_dir: Optional[Path],
+    hdd: bool,
+    check_only: bool,
+):
```
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py:79]: yes, this is checking for available disk space for 

- temporary tables and
- the whole database 

I am assuming that we have sufficient free space when the size of the whole v1 database is < available free space.
Temp space is only reported unless a separate temp directory is chosen. Then I also check the target path for available free space. The reason is, that when we always check for 2*v1 db size, then we will never exceed this limit with temp space

this calculates the db size: `db_size_byte = page_count[0] * page_size[0]`

File: chia/cmds/db_upgrade_alt_func.py [hunk 3]
```diff
@@ -0,0 +1,850 @@
...
+                f"\n\tmissing {db_size_byte-db_store_free} byte/"
+                f"{round(db_size_byte-db_store_free,2)/1024/1024/1024} GiB"
+            )
+            failed = True
+        if failed:
+            raise RuntimeError("checks failed.")
+        else:
+            print("checks passed. proceeding with upgrade")
+    except sqlite3.OperationalError:
+        pass
```
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py:139]: probably not. I think I just took this part from [here](https://github.com/Chia-Network/chia-blockchain/blob/419f88f296b3e61d3af8fc070f7b31546f55bbe1/chia/cmds/db_upgrade_func.py#L83) without thinking much about it.

File: chia/cmds/db_upgrade_alt_func.py [hunk 4]
```diff
@@ -0,0 +1,850 @@
...
+        if row is not None and row[0] == 2:
+            print(f"blockchain database already version {row[0]}\nDone")
+            raise RuntimeError("already v2")
+    except sqlite3.OperationalError:
+        pass
+    finally:
+        cursor.close()
+
+    if offline:
+        db.execute("PRAGMA JOURNAL_MODE=off")
```
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py:167]: For the output database it is set in line 188 directly.
For the input database, I think the performance improvement using no journal is the reduced overhead. When you compare my test runtimes you see, that using my High-end HW there is almost no difference in runtime, but for the RPi it is a reduction of 17m.

If I would have only one production full node and a couple of other nodes in my environment, than I would always opt to use the offline mode for all the other db upgrades.

File: chia/cmds/db_upgrade_alt_func.py [hunk 5]
```diff
@@ -0,0 +1,850 @@
...
+
+    in_db.close()
+    ############
+
+    # Reconnect to get new mmap initalization for new table
+    # Change temp file path if specified
+    if temp_store_path is not None:
+        temp2_out_path = temp_store_path / "temp_coin_store.sqlite"
+    else:
+        temp2_out_path = out_path.parent / "temp_coin_store.sqlite"
```
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py:515]: I tried to explain it [here](https://github.com/neurosis69/chia-blockchain/blob/3ad841902c8d699db06acf6c121dd77842417ae8/chia/cmds/db_upgrade_alt_func.py#L342) in the comment.

Let me know if I should get into more details.

File: chia/cmds/db_upgrade_alt_func.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (arvidn) [reply chia/cmds/db_upgrade_alt_func.py:167]: so, you're saying this *does* affect queries too then? a query wouldn't need to write any journal either way, but perhaps this also means "and we can assume there are no other connections, to this db, configured to use a journal".

If setting this alone on the input database saves 17m, it seems like an obvious option to add.

File: chia/cmds/db_upgrade_alt_func.py [hunk 6]
```diff
@@ -0,0 +1,850 @@
...
+        for row in rows:
+            # The following part was in the original upgrade script
+            # Not sure if this was necessary,
+            # as the fetching query already had a condition where confirmed_index <= peak
+
+            # in order to convert a consistent snapshot of the
+            # blockchain state, any coin that was spent *after* our
+            # cutoff must be converted into an unspent coin
+            # if spent_index > peak_height:
+            #    spent_index = 0
```
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py:593]: ah, ok . I think I got it now.
this is than of course necessary and I will add it back.

File: chia/cmds/db_upgrade_alt_func.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py:167]: well, I just took my time and read chapter 2.2 + 2.3. [here](https://www.sqlite.org/wal.html) and about the isolation [here](https://www.sqlite.org/isolation.html)

How I understand this is, that when reading we 

- need to check the WAL file if there is data we need for our select
- prevent checkpointing progress for very long selects

Maybe even more. 
I meant, having this only as operational overhead could affect migrations on RPi or even HDDs.

**Comment by neurosis69** [comment]
Regarding consistency, if we want a consistent state we could also try to use `BEGIN TRANSACTION` before copying the date of all full_blocks, block_records and coin_records

File: chia/cmds/db_upgrade_alt_func.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (neurosis69) [reply chia/cmds/db_upgrade_alt_func.py]: added with [1c962ef](https://github.com/Chia-Network/chia-blockchain/pull/10116/commits/1c962ef8cb07e32ccedafa71aec8ac4fe6f29c97)

**Comment by arvidn** [comment]
from your benchmarks, it doesn't look like the `offline` parameter has that big of an impact. ~20 seconds out of 16 minutes. It's a very simple patch though.

I imagine the main improvement is using a temporary table that the database can be vacuumed into. Did you consider vacuuming the source database in-place instead?

with an offline option, the input database could be modified (i.e. vacuumed) which could provide much simpler logic and perhaps give similar speedups.

**Comment by neurosis69** [comment]
And I think it is easier for an enduser to just execute `chia db upgrade --offline` and the script takes care of anyhting else, than asking someone to do vacuum in sqlite themselves.

**Comment by arvidn** [comment]
I was thinking of something like this: https://github.com/Chia-Network/chia-blockchain/pull/10193

**Comment by neurosis69** [comment]
I'll check #10193 in my evening in a couple of hours.

But I just remembered. If you want to use `vacuum` for simplicity than you can also offer this for online dbs using `vacuum into`. This is transaction safe and I and my pool members have done this dozens of times as daily backup .. and of course, some of us had the need to restore from this "backup". (which brings me to the idea adding a function `chia db backup` using this command).

**Pull request closed without merging**